File: api-java/src/main/java/fr/huiitre/tools/modules/riot/valorant/application/catalog/usecase/GetValorantBundleUseCase.java
package fr.huiitre.tools.modules.riot.valorant.application.catalog.usecase;

import java.util.Optional;

import org.springframework.stereotype.Service;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.ports.ValorantBundleRepository;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.view.ValorantBundleView;

@Service
public class GetValorantBundleUseCase implements SecuredUseCase {

    private final ValorantBundleRepository bundleRepository;

    public GetValorantBundleUseCase(ValorantBundleRepository bundleRepository) {
        this.bundleRepository = bundleRepository;
    }

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.READ_ONLY;
    }

    public ValorantBundleView execute(Long id) {
        return bundleRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Bundle not found: " + id));
    }
}

File: api/Modules/Riot/Valorant/Application/Catalog/Usecases/GetValorantBundleUseCase.cs
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Views;

namespace Tools.Api.Modules.Riot.Valorant.Application.Catalog.Usecases;

public sealed class GetValorantBundleUseCase(
    UseCaseAuthorizer authorizer,
    IValorantBundleRepository bundleRepository
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.ReadOnly;
    protected override ModuleCode? RequiredModule => ModuleCode.Riot;

    public async Task<ValorantBundleView> Execute(long id)
    {
        var bundle = await bundleRepository.FindById(id);

        return bundle ?? throw AppException.NotFound(
            "VALORANT_BUNDLE_NOT_FOUND",
            $"Le pack {id} est introuvable.");
    }
}
